Add preprocessors `distance_metrics` and `histogram` (#2299)
ESMValGroup/ESMValCore: doc/recipe/preprocessor.rst

recipe:

.. _list_of_distance_metrics:

* ``metric`` (:obj:`str`): Distance metric that is calculated.
  Must be one of

  * ``'rmse'``: `Unweighted root mean square error`_.

  .. math::

    RMSE = \sqrt{\frac{1}{N} \sum_{i=1}^N \left( x_i - r_i \right)^2}

  * ``'weighted_rmse'``: `Weighted root mean square error`_.

  .. math::

    WRMSE = \sqrt{\sum_{i=1}^N w_i \left( x_i - r_i \right)^2}

  * ``'pearsonr'``: `Unweighted Pearson correlation coefficient`_.

  .. math::

    r = \frac{
      \sum_{i=1}^N
      \left( x_i - \bar{x} \right) \left( r_i - \bar{r} \right)
    }{
      \sqrt{\sum_{i=1}^N \left( x_i - \bar{x} \right)^2}
      \sqrt{\sum_{i=1}^N \left( r_i - \bar{r} \right)^2}
    }

  * ``'weighted_pearsonr'``: `Weighted Pearson correlation coefficient`_.

  .. math::

    r = \frac{
      \sum_{i=1}^N
      w_i \left( x_i - \bar{x} \right) \left( r_i - \bar{r} \right)
    }{
      \sqrt{\sum_{i=1}^N w_i \left( x_i - \bar{x} \right)^2}
      \sqrt{\sum_{i=1}^N w_i \left( r_i - \bar{r} \right)^2}
    }

  * ``'emd'``: `Unweighted Earth mover's distance`_ (EMD).
    The EMD is also known as first Wasserstein metric `W`\ :sub:`1`, which is a

>     metric that measures distances between two probability distributions.

Reviewer: ```suggestion
    metric that measures distance between two probability distributions.
```

    For this, discrete probability distributions of the input data are created
    through binning, which are then used as input for the Wasserstein metric.
    The metric is also known as `Earth mover's distance` since, intuitively, it
    can be seen as the minimum "cost" of turning one pile of earth into another
    one (pile of earth = probability distribution).
    This is also known as `optimal transport` problem.
    Formally, this can be described with a joint probability distribution (or
    `optimal transport matrix`) γ (whose marginals are the input distributions)
    that minimizes the "transportation cost":

  .. math::

    W_1 = \min_{\gamma \in \mathbb{R}^{n \times n}_{+}} \sum_{i,j}^{n}
    \gamma_{ij} \lvert X_i - R_i \rvert \\
    \textrm{with} ~~ \gamma 1 = p_X(X);~ \gamma^T 1 = p_R(R)

  * ``'weighted_emd'``: `Weighted Earth mover's distance`_.
    Similar to the unweighted EMD (see above), but here weights are considered
    when calculating the probability distributions (i.e., instead of 1, each

>     element contributes a given weight in the bin count; see also `weights`

Reviewer: ```suggestion
    element provides a weight in the bin count; see also `weights`
```

      clip:
        minimum: 0
        maximum: null

.. _histogram:

``histogram``
-------------------

This function calculates histograms.

The ``histogram`` preprocessor supports the following arguments in the
recipe:

* ``coords`` (:obj:`list` of :obj:`str`, default: ``None``): Coordinates over
  which the histogram is calculated.
  If ``None``, calculate the histogram over all coordinates.
  The shape of the output cube will be `(x1, x2, ..., n_bins)`, where `xi` are
  the dimensions of the input cube not appearing in `coords` and `n_bins` is
  the number of bins.
* ``bins`` (:obj:`int` or sequence of :obj:`float`, default: 10): If `bins` is
  an :obj:`int`, it defines the number of equal-width bins in the given
  `bin_range`.
  If `bins` is a sequence, it defines a monotonically increasing array of bin
  edges, including the rightmost edge, allowing for non-uniform bin widths.

>   also calculate the distance of the reference dataset with itself.

Reviewer: ```suggestion
```

* ``bin_range`` (:obj:`tuple` of :obj:`float` or ``None``, default: ``None``):
  The lower and upper range of the bins.
  If ``None``, `bin_range` is simply (``cube.core_data().min(),
  cube.core_data().max()``).
  Values outside the range are ignored.
  The first element of the range must be less than or equal to the second.
  `bin_range` affects the automatic bin computation as well if `bins` is an
  :obj:`int` (see description for `bins` above).
* ``weights`` (array-like, :obj:`bool`, or ``None``, default: ``None``):
  Weights for the histogram calculation.
  Each value in the input data only contributes its associated weight towards
  the bin count (instead of 1).
  Weights are normalized before entering the calculation if `normalization` is
  ``'integral'`` or ``'sum'``.
  Can be an array of the same shape as the input data, ``False`` or ``None``
  (no weighting), or ``True``.
  In the latter case, weighting will depend on `coords`, and the following
  coordinates will trigger weighting: `time` (will use lengths of time
  intervals as weights) and `latitude` (will use cell area weights).
  Time weights are always calculated from the input data.

>   Area weights can be given as supplementary variables to the recipe

Reviewer: ```suggestion
  Area weights can be given as supplementary variables in the recipe
```

  (`areacella` or `areacello`, see :ref:`supplementary_variables`) or
  calculated from the input data (this only works for regular grids).
  By default, **NO** supplementary variables will be used; they need to be
  explicitly requested in the recipe.
* ``normalization`` (``None``, ``'sum'``, or ``'integral'``, default:
  ``None``): If ``None``, the result will contain the number of samples in each
  bin.
  If ``'integral'``, the result is the value of the probability `density`
  function at the bin, normalized such that the integral over the range is 1.
  If ``'sum'``, the result is the value of the probability `mass` function at

>   the bin, normalized such that the sum over the range is 1.

Reviewer: ```suggestion
  the bin, normalized such that the sum over the whole range is 1.
```
